File: src/main/java/me/chell/samsara/impl/module/misc/KillEffects.java
package me.chell.samsara.impl.module.misc;

import com.google.common.collect.Lists;
import me.chell.samsara.Samsara;
import me.chell.samsara.api.event.PlayerKilledEvent;
import me.chell.samsara.api.event.PlayerUpdateEvent;
import me.chell.samsara.api.module.Module;
import me.chell.samsara.api.util.ChatUtil;
import me.chell.samsara.api.value.Value;
import net.minecraft.entity.effect.EntityLightningBolt;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.SoundEvents;
import net.minecraft.util.EnumParticleTypes;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.SoundEvent;
import net.minecraftforge.event.world.WorldEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.registry.ForgeRegistries;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

public class KillEffects extends Module {

    public KillEffects() {
        super("KillEffects", Category.MISC);
        sounds = Lists.newArrayList(
                registerSound("godlike"),
                registerSound("holyshit"),
                registerSound("impressive"),
                registerSound("ownage"),
                registerSound("perfect"),
                registerSound("wickedsick")
        );
        playerMap = new HashMap<>();
        rand = new Random();
    }

    private final Value<QuakeMode> quake = builder("Announcer", QuakeMode.OFF).build(); // Doesn't work in dev environment because idk how to do the sounds.json for assets/minecraft
    private final Value<ChatMode> chat = builder("Chat", ChatMode.OFF).build();
    private final Value<String> message = builder("Message", "GG ${player}, well played!").visible(b -> chat.getValue() != ChatMode.OFF).build();
    private final Value<ParticleMode> particles = builder("Particles", ParticleMode.OFF).build();
    private final Value<Boolean> particleSound = builder("Sound", true).visible(b -> particles.getValue() == ParticleMode.THUNDER).build();
    private final Value<Float> heartsTime = builder("Time", 2f).bounds(0f, 5f).visible(b -> particles.getValue() == ParticleMode.HEARTS).build();
    private final Value<Integer> heartsAmount = builder("Amount", 8).bounds(1, 10).visible(b -> particles.getValue() == ParticleMode.HEARTS).build();

    private final List<SoundEvent> sounds;
    private int kills;
    private final HashMap<EntityPlayer, Integer> playerMap;
    private final Random rand;

    @SubscribeEvent
    public void onPlayerKilled(PlayerKilledEvent event) {
        kills++;
        sendMessage(event.getTarget());
        doParticleEffect(event.getTarget());
        playSound();
    }

    @SubscribeEvent
    public void onWorldLoaded(WorldEvent.Load event) {
        kills = 0;
        playerMap.clear();
    }

    @SubscribeEvent
    public void onPlayerUpdate(PlayerUpdateEvent event) {
        // net.minecraft.entity.passive.EntityAnimal
        List<EntityPlayer> remove = new ArrayList<>();

        for(EntityPlayer player : playerMap.keySet()) {
            int particleTime = playerMap.get(player);
            if(particleTime > 0) {
                playerMap.put(player, particleTime-1);

                if (particleTime % (11-heartsAmount.getValue()) == 0) {
                    double d0 = rand.nextGaussian() * 0.02D;
                    double d1 = rand.nextGaussian() * 0.02D;
                    double d2 = rand.nextGaussian() * 0.02D;
                    getWorld().spawnParticle(EnumParticleTypes.HEART,
                            player.posX + (double)(rand.nextFloat() * player.width * 4.0F) - (double)player.width,
                            player.posY + 0.5D + (double)(rand.nextFloat() * player.height),
                            player.posZ + (double)(rand.nextFloat() * player.width * 4.0F) - (double)player.width,
                            d0, d1, d2);
                }
            } else {
                remove.add(player);
            }
        }

        for(EntityPlayer player : remove) {
            playerMap.remove(player);
        }
    }

    private void sendMessage(EntityPlayer target) {
        if(chat.getValue().equals(ChatMode.CLIENT)) {
            ChatUtil.sendClientMessage(message.getValue(), ChatUtil.var("player", target.getName()));
        } else if(chat.getValue().equals(ChatMode.ON)) {
            ChatUtil.sendChatMessage(message.getValue(), ChatUtil.var("player", target.getName()));
        }
    }

    private void playSound() {
        if(quake.getValue().equals(QuakeMode.RANDOM)) {
            int rand = ThreadLocalRandom.current().nextInt(0, sounds.size());
            getPlayer().playSound(sounds.get(rand), 1f, 1f);
        }
    }

    private void doParticleEffect(EntityPlayer target) {
        switch (particles.getValue()) {
            case THUNDER:
                getWorld().addWeatherEffect(new EntityLightningBolt(getWorld(), target.posX, target.posY, target.posZ, true));
                if(particleSound.getValue()) {
                    getWorld().playSound(getPlayer(), target.posX, target.posY, target.posZ, SoundEvents.ENTITY_LIGHTNING_THUNDER, SoundCategory.WEATHER, 10000.0F, 0.9F);
                    getWorld().playSound(getPlayer(), target.posX, target.posY, target.posZ, SoundEvents.ENTITY_LIGHTNING_IMPACT, SoundCategory.WEATHER, 2.0F, 0.6F);
                }
                break;
            case HEARTS:
                playerMap.put(target, (int)(heartsTime.getValue() * 20));
                break;
        }
    }

    private SoundEvent registerSound(String name){
        SoundEvent soundEvent = new SoundEvent(new ResourceLocation(Samsara.MODID, name));
        soundEvent.setRegistryName(name);
        ForgeRegistries.SOUND_EVENTS.register(soundEvent);
        return soundEvent;
    }

    private enum QuakeMode {
        OFF, RANDOM, MULTI // TODO double kill, triple kill etc. I can't find the proper sound effects anywhere
    }

    private enum ChatMode {
        OFF, ON, CLIENT
    }

    private enum ParticleMode {
        OFF, THUNDER, HEARTS
    }
}

File: src/main/kotlin/me/chell/samsara/impl/module/combat/KillEffects.kt
package me.chell.samsara.impl.module.combat

import me.chell.samsara.api.event.EventHandler
import me.chell.samsara.api.module.Module
import me.chell.samsara.api.value.Register
import me.chell.samsara.api.value.Value
import me.chell.samsara.impl.event.EntityKilledEvent
import me.chell.samsara.impl.event.PlayerTickEvent
import net.minecraft.entity.EntityType
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.mob.HostileEntity
import net.minecraft.entity.passive.PassiveEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.particle.ParticleTypes
import net.minecraft.sound.SoundEvent
import net.minecraft.text.LiteralText
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import kotlin.random.Random

class KillEffects: Module("KillEffects", Category.COMBAT) {

    private val sounds = mutableListOf<SoundEvent>()

    init {
        sounds.add(registerSoundEvent("godlike"))
        sounds.add(registerSoundEvent("holyshit"))
        sounds.add(registerSoundEvent("impressive"))
        sounds.add(registerSoundEvent("ownage"))
        sounds.add(registerSoundEvent("perfect"))
        sounds.add(registerSoundEvent("wickedsick"))
    }

    @Register(0) val quake = Value("Announcer", true)

    @Register(1) val chat = Value("Chat", Chat.CLIENT)
    @Register(2) val message = Value("Message", "GG, \${target}, well played!", visible = { chat.value != Chat.OFF })

    @Register(3) val particles = Value("Particles", Particles.OFF)
    @Register(4) val amount = Value("Amount", 8, min = 1, max = 10, visible = { particles.value != Particles.OFF })
    @Register(5) val time = Value("Time", 2.0, min = 0.0, max = 5.0, visible = { particles.value == Particles.HEARTS })

    @Register(6) val players = Value("Players", true)
    @Register(7) val hostile = Value("Hostile", false)
    @Register(8) val passive = Value("Passive", false)

    private val entityMap = mutableMapOf<LivingEntity, Int>()

    @EventHandler
    fun onKill(event: EntityKilledEvent) {
        sendMessage(event.target)
        spawnParticles(event.target)
        playSound()
    }

    @EventHandler
    fun onPlayerTick(event: PlayerTickEvent) {
        val list = mutableListOf<LivingEntity>()

        for(entity in entityMap.keys.toList()) {
            val t = entityMap[entity]!!
            if(t > 0) {
                entityMap[entity] = t-1

                if(t % (11-amount.value) == 0) {
                    for(i in 0..2) {
                        val d0 = Random.nextDouble() * 0.02
                        val d1 = Random.nextDouble() * 0.02
                        val d2 = Random.nextDouble() * 0.02

                        world.addParticle(
                            ParticleTypes.HEART,
                            entity.x + Random.nextDouble() * entity.width * 4.0 - entity.width,
                            entity.y + 0.5 + Random.nextDouble() * entity.height,
                            entity.z + Random.nextDouble() * entity.width * 4.0 - entity.width,
                            d0, d1, d2
                        )
                    }
                }
            } else {
                list.add(entity)
            }
        }

        for(entity in list) {
            entityMap.remove(entity)
        }
    }

    private fun sendMessage(target: LivingEntity) { // TODO ChatManager to avoid spam
        val msg = message.value.replace("\${target}", target.name.string)
        when(chat.value) {
            Chat.SERVER -> {
                player.sendChatMessage(msg)
            }
            Chat.CLIENT -> {
                player.sendMessage(LiteralText(msg), false)
            }
            else -> {}
        }
    }

    private fun playSound() {
        if(quake.value)
            player.playSound(sounds[Random.nextInt(sounds.size)], 1f, 1f)
    }

    private fun spawnParticles(target: LivingEntity) {
        when(particles.value) {
            Particles.LIGHTNING -> {
                for(i in 0 until amount.value) {
                    val entity = EntityType.LIGHTNING_BOLT.create(world)!!
                    entity.setCosmetic(true)
                    entity.refreshPositionAfterTeleport(target.pos)
                    world.addEntity(6342792+i, entity) // random number hopefully high enough to not delete any entities
                }

            }
            Particles.HEARTS -> {
                entityMap[target] = (time.value * 20).toInt()
            }
            else -> {}
        }
    }

    private fun registerSoundEvent(name: String): SoundEvent {
        val id = Identifier("$MODID:$name")
        val event = SoundEvent(id)
        Registry.register(Registry.SOUND_EVENT, id, event)
        return event
    }

    private fun isValid(entity: LivingEntity): Boolean {
        return (entity is PlayerEntity && players.value)
                || (entity is HostileEntity && hostile.value)
                || (entity is PassiveEntity && passive.value)
    }

    enum class Chat {
        OFF, SERVER, CLIENT
    }

    enum class Particles {
        OFF, LIGHTNING, HEARTS
    }

}